src/client/eos-client.ts

```diff
@@ -37,30 +38,79 @@ export class EOSClient {
     return this.client.connect(onMessage)
   }
 
+  public send<T extends OutboundMessage<{}>>(
+    type: OutboundMessageType,
+    messageOptions: T,
+    ...messageTypes: InboundMessageType[]
+  ) {
+    return this.sendAndListen<T>(
+      messageOptions,
+      messageOptions.req_id === undefined ? type : messageOptions.req_id,
+      messageOptions.listen === undefined ? true : messageOptions.listen,
+      ...messageTypes
+    )
+  }
+
   public getActions(parameters: GetActionsMessageParameters, options: StreamOptions = {}) {
     options = withDefaults({ listen: true }, options)
-    this.client.send(getActionsMessage(parameters, options))
-
-    return this.createListener(options.requestId!, InboundMessageType.ACTION_TRACE)
+    const messageOptions = getActionsMessage(parameters, options)
+
+    return this.sendAndListen<OutboundMessage<GetActionsMessageBackendParameters>>(
+      messageOptions,
+      options.requestId!,
+      options.listen!,
+      InboundMessageType.ACTION_TRACE
+    )
   }
 
   public getTableRows(parameters: GetTableRowsMessageParameters, options: StreamOptions = {}) {
     options = withDefaults({ listen: true }, options)
-    this.client.send(getTableRowsMessage(parameters, options))
-
-    return this.createListener(options.requestId!, InboundMessageType.TABLE_DELTA)
+    const messageOptions = getTableRowsMessage(parameters, options)
+    return this.sendAndListen<OutboundMessage<GetTableRowsMessageBackendParameters>>(
+      messageOptions,
+      options.requestId!,
+      options.listen!,
+      InboundMessageType.TABLE_DELTA
+    )
   }
 
   public getTransaction(id: string, options: StreamOptions = {}) {
     options = withDefaults({ fetch: true, listen: true }, options)
-    this.client.send(getTransactionMessage(id, options))
+    const messageOptions = getTransactionMessage(id, options)
+
+    return this.sendAndListen<OutboundMessage<{ id: string }>>(
+      messageOptions,
```

Comment:
Those three lines are repeated for each implementation, we could just send the message as-is and the consumer could extract the required information from it.